src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+                    && isExpired(oldestMembership.gracePeriodStartDate)
+            ) {
+                emit MemberExpired(head, oldestMembership.userMessageLimit, oldestMembership.index);
+
+                // Deduct the expired membership rate limit
+                totalRateLimitPerEpoch -= oldestMembership.userMessageLimit;
+
+                // Promote the next oldest membership to oldest
+                uint256 nextOldest = oldestMembership.next;
+                head = nextOldest;
+                if (nextOldest != 0) {
+                    members[nextOldest].prev = 0;
+                }
+
+                // Move balance from expired membership to holder balance
+                balancesToWithdraw[oldestMembership.holder][oldestMembership.token] += oldestMembership.amount;
+
+                availableExpiredIndices.push(oldestMembership.index);
+
+                delete members[head];
+            } else {
+                revert ExceedMaxRateLimitPerEpoch();
+            }
+        }
+
+        uint256 prev = 0;
+        if (tail != 0) {
+            MembershipInfo storage latestMembership = members[tail];
+            latestMembership.next = _idCommitment;
+            prev = tail;
+        } else {
+            // First item
+            // TODO: test adding memberships after the list has been emptied
+            head = _idCommitment;
+        }
+
+        totalRateLimitPerEpoch += _rateLimit;
+
+        // Reuse available slots from previously removed expired memberships
+        uint256 arrLen = availableExpiredIndices.length;
+        if (arrLen != 0) {
+            index = availableExpiredIndices[arrLen - 1];
+            availableExpiredIndices.pop();
+            reusedIndex = true;
+        } else {
+            index = commitmentIndex;
+        }
+
+        members[_idCommitment] = MembershipInfo({
+            holder: _sender,
+            gracePeriodStartDate: block.timestamp + (uint256(billingPeriod) * uint256(_numberOfPeriods)),
+            gracePeriod: gracePeriod,
+            numberOfPeriods: _numberOfPeriods,
+            token: _token,
+            amount: _amount,
+            userMessageLimit: _rateLimit,
+            next: 0, // It's the last value, so point to nowhere
+            prev: prev,
+            index: index
+        });
+
+        tail = _idCommitment;
+    }
+
+    /// @dev Extend a membership expiration date. Membership must be on grace period
+    /// @param _sender the address of the holder of the membership
+    /// @param _idCommitment the idCommitment of the membership
+    function _extendMembership(address _sender, uint256 _idCommitment) public {
+        MembershipInfo storage mdetails = members[_idCommitment];
+
+        if (!_isGracePeriod(mdetails.gracePeriodStartDate, mdetails.gracePeriod, mdetails.numberOfPeriods)) {
+            // TODO: can a membership that has exceeded the expired period be extended?
```

Comment:
?

Reply:
I'd say no: extensions are only possible from _GracePeriod_ as per spec.

However, I'm not sure I inderstand what "exceeded the expired period" means. When the membership is in _Expired_ state, it stays there forever unless either the holder withdraws the deposit, or another membership overwrites the slot... Or what does "expired period" mean here?

Reply:
> I'd say no: extensions are only possible from GracePeriod as per spec.

any reason why imposing this limit? I would say allowing to renew/extend at any time would be desirable.

Reply:
> any reason why imposing this limit? I would say allowing to renew/extend at any time would be desirable.

The spec FAQ says:

> The rationale is to make possible parameter changes that the contract _Owner_ might make (e.g., for security reasons) applicable to most memberships.

In other words, we want to limit extensions somewhat to pervent users from keeping old membership parameters for too long if the contract Owner changes them.

However, I feel that it isn't a really solid justification. Strictly speaking, there are multiple design goals that we can aim for in parallel:

1. make the state transition diagram as simple as possible;
2. make the UX familiar and friendly;
3. prevent users from avoiding parameter changes that the contract Owner might introduce.

Only allowing extensions from GracePeriod helps (1), hurts (2), and... is neutral towards (3) (though I'm not sure about the last statement).

Reply:
> In other words, we want to limit extensions somewhat to pervent users from keeping old membership parameters for too long if the contract Owner changes them.

This makes sense, but we are talking about two different things? I mean you can allow extensions at any time (not just from grace period) but also limit it to X to prevent "old membership parameters for too long if the contract Owner changes them."

Reply:
I mean you can allow extending at any time as long as the total remaining time is not more than x days.

Reply:
Created an [issue](https://github.com/waku-org/specs/issues/38) to discuss this matter further. As decided on the research call (2024-09-09), these potential changes should be considered separately for a future iteration of the spec.

Reply:
Removed the `TODO` from code in https://github.com/waku-org/waku-rlnv2-contract/commit/0cd80d639c1f6b3b0b8311dc6f1c1777e9a36323